### PR TITLE
Mitigate protected-access error by encapsulating factory checks

### PR DESCRIPTION
**Pull request by ankona**
Shift checks for a managed indicator into external code.

### CODE REVIEW

**Review by MattToast — COMMENTED** [review]
Interesting approach! I have a couple of preliminary thoughts (some of which conflicting, lol). Let me know what you think!

File: src/python/module/smartredis/configoptions.py [hunk 1]
```diff
@@ -25,20 +25,55 @@
...
+    @classmethod
+    def error_message(mcs, class_name: str) -> str:
+        return mcs.NOT_MANAGED.format(class_name)
+
+    @classmethod
+    def construct_class(mcs, bases: t.Tuple[t.Type]) -> t.Any:
+        """Factory method for creating managed instances"""
+        bases = (_Managed,) + bases
+        class_name = "".join(base.__name__ for base in bases)
+        managed_class = _FactoryMeta(class_name, bases, {})
```
[the remark this answers is not in the document]
Comment (MattToast) [reply src/python/module/smartredis/configoptions.py]: Fixed in commit sense this was written and no longer relevant!

File: src/python/module/smartredis/configoptions.py [hunk 2]
```diff
@@ -25,20 +25,55 @@
...
+class _Managed:
+    def __init__(self) -> None:
+        self._is_managed = True
+
+class _FactoryMeta(type):
+    NOT_MANAGED = "Managed method called from an unmanaged {0} instance"
+
+    def __new__(mcs, name: str, bases: t.Tuple[t.Type], attrs: t.Dict) -> t.Type:
+        """Instantiate and mark an object as managed"""
+        return super().__new__(mcs, name, bases, attrs)
```
[the remark this answers is not in the document]
Comment (ankona) [reply src/python/module/smartredis/configoptions.py:44]: yes, yes!

File: src/python/module/smartredis/configoptions.py [hunk 3]
```diff
@@ -25,20 +25,55 @@
...
 from .smartredisPy import PyConfigOptions
 from .util import exception_handler, typecheck
 
-_NOT_FACTORY = (
-    "Method called on a ConfigOptions object not created from a factory method"
-)
+
+class _Managed:
+    def __init__(self) -> None:
+        self._is_managed = True
```
[the remark this answers is not in the document]
Comment (ankona) [reply src/python/module/smartredis/configoptions.py:37]: Great eye. Added.

File: src/python/module/smartredis/configoptions.py [hunk 4]
```diff
@@ -25,20 +25,55 @@
...
+    def _wrapper(*args: t.Any, **kwargs: t.Any) -> t.Any:
+        _FactoryMeta.verify_managed(args[0])
+        return func(*args, **kwargs)
+    return _wrapper
 
 
 class ConfigOptions:
     def __init__(self) -> None:
         """Initialize a ConfigOptions base object"""
-        self._is_created_via_factory = False
```
[the remark this answers is not in the document]
Comment (ankona) [reply src/python/module/smartredis/configoptions.py:41]: It looks like the goal was to encapsulate the usage of pybind under the hood creating the PyConfigOptions. @billschereriii can probably explain the original thought process.

File: src/python/module/smartredis/configoptions.py [hunk 5]
```diff
@@ -52,9 +87,9 @@ def from_pybind(configoptions: PyConfigOptions) -> "ConfigOptions":
         :rtype: ConfigOptions
         """
         typecheck(configoptions, "configoptions", PyConfigOptions)
-        new_configoptions = ConfigOptions()
-        new_configoptions.set_configoptions(configoptions)
-        return new_configoptions
+        opts: ConfigOptions = _FactoryMeta.construct_class(bases=(ConfigOptions,))
```
[the remark this answers is not in the document]
Comment (ankona) [reply src/python/module/smartredis/configoptions.py:87]: I actually liked the idea that the Factory knows how to create some object and only the factory knows how to tell you if it came from its "assembly line." BUT, from a practical perspective, I can now see that I can get rid of `_verify_managed` altogether moving a simple `isinstance(args[0], _Managed)` into the decorator. And moving the error message to be a byproduct of the decorator also makes sense (why does the factory need to know how we talk about it).

I killed both extraneous methods! Great thinking.

File: src/python/module/smartredis/configoptions.py [hunk 6]
```diff
@@ -25,20 +25,55 @@
...
+    def error_message(mcs, class_name: str) -> str:
+        return mcs.NOT_MANAGED.format(class_name)
+
+    @classmethod
+    def construct_class(mcs, bases: t.Tuple[t.Type]) -> t.Any:
+        """Factory method for creating managed instances"""
+        bases = (_Managed,) + bases
+        class_name = "".join(base.__name__ for base in bases)
+        managed_class = _FactoryMeta(class_name, bases, {})
+        return managed_class()
```
[the remark this answers is not in the document]
Comment (ankona) [reply src/python/module/smartredis/configoptions.py:46]: Long live the Liskov Substitution Principle!

[the remark this answers is not in the document]
Comment (ankona) [reply src/python/module/smartredis/configoptions.py:47]: Added a bit of uuid magic to avoid the appearance of the same things.

File: tests/python/test_errors.py [hunk 7]
```diff
@@ -32,6 +32,12 @@
 from smartredis.error import *
 
 
+@pytest.fixture
+def cfg_opts() -> ConfigOptions:
+    opts = ConfigOptions.create_from_environment("")
+    return opts
```
Comment (billschereriii) [inline tests/python/test_errors.py:38]: Is this even used?

File: tests/python/test_errors.py [hunk 8]
```diff
@@ -913,43 +919,53 @@ def test_get_tensor_names_wrong_type():
...
-    co = ConfigOptions()
-    key = "stringval"
-    value = "target_string_value"
+@pytest.mark.parametrize(
+        "key,value",
+        [
+            pytest.param(42, 42, id="Invalid key type"),
+            pytest.param("key", "stringval", id="Invalid value type"),
+        ]
+)
```
Comment (billschereriii) [inline tests/python/test_errors.py:950]: This parameterization obscures the testing done here (and differs from all the other testing in this file). Please revert this change

File: tests/python/test_errors.py [hunk 9]
```diff
@@ -913,43 +919,53 @@ def test_get_tensor_names_wrong_type():
...
-        _ = co.override_string_option(42, value)
+        _ = cfg_opts.override_integer_option(key, value)
+
+@pytest.mark.parametrize(
+        "key,value",
+        [
+            pytest.param(42, "stringval", id="Invalid key type"),
+            pytest.param("stringval", 42, id="Invalid value type"),
+        ]
+)
```
Comment (billschereriii) [inline tests/python/test_errors.py:963]: This parameterization obscures the testing done here (and differs from all the other testing in this file). Please revert this change

File: src/python/module/smartredis/configoptions.py [hunk 10]
```diff
@@ -25,20 +25,45 @@
...
+        bases = (_Managed, base)
+        unique_key = str(uuid4()).split("-", 1)[0]
+        class_name = "".join(base.__name__ for base in bases) + unique_key
+        managed_class = mcs(class_name, bases, {})
+        return managed_class()
+
+
+def managed(func: t.Callable) -> t.Callable:
+    """Decorator to verify that a class was constructed using a factory"""
+    not_managed = "Managed method called from an unmanaged {0} instance"
```
Comment (billschereriii) [inline src/python/module/smartredis/configoptions.py:52]: This error message is not useful for a user as it doesn't explain what they did wrong to get here. Please revert to the original error message

File: src/python/module/smartredis/configoptions.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (billschereriii) [reply src/python/module/smartredis/configoptions.py:41]: The goal here is to catch the case where users try to instantiate an instance of the class directly and work with it. Python lacks protections to block direct use of __init__() so we had to work around that. Only an instance created via create_from_environment() (or another -- future -- factory constructor can be used.

File: tests/python/test_errors.py (hunk 8 above)
Comment (ankona) [reply tests/python/test_errors.py:950]: Done!

File: tests/python/test_errors.py (hunk 9 above)
Comment (ankona) [reply tests/python/test_errors.py:963]: Done!

File: src/python/module/smartredis/configoptions.py [hunk 11]
```diff
@@ -25,20 +25,48 @@
...
+
+class _Managed:
+    """Marker class identifying factory-created objects"""
+
+
+def create_config_options(base: t.Type[t.Any]) -> t.Any:
+    """Factory method for creating managed instances"""
+    bases = (_Managed, base)
+    unique_key = str(uuid4()).split("-", 1)[0]
+    class_name = "".join(base.__name__ for base in bases) + unique_key
```
Comment (billschereriii) [inline src/python/module/smartredis/configoptions.py:44]: I'd still really like some extra documentation on these two lines for if I have to come back and do maintenance on it in 6 months

File: src/python/module/smartredis/configoptions.py [hunk 12]
```diff
@@ -25,20 +25,45 @@
...
+
+class _FactoryMeta(type):
+    @classmethod
+    def instantiate(mcs, base: t.Type[t.Any]) -> t.Any:
+        """Factory method for creating managed instances"""
+        bases = (_Managed, base)
+        unique_key = str(uuid4()).split("-", 1)[0]
+        class_name = "".join(base.__name__ for base in bases) + unique_key
+        managed_class = mcs(class_name, bases, {})
+        return managed_class()
```
[the remark this answers is not in the document]
Comment (ankona) [reply src/python/module/smartredis/configoptions.py]: Shifted the name generation into nested method so it could have a nice docstring & name to make that part clear.

Added a comment to explain the base class tuple

File: src/python/module/smartredis/configoptions.py (hunk 11 above)
Comment (ankona) [reply src/python/module/smartredis/configoptions.py]: please see added comments & nested method

**Review by billschereriii — APPROVED** [review]
Looks great. Thanks for working through this one!
